packages/api/amplify_api/lib/src/decorators/authorize_websocket_message.dart

```diff
@@ -0,0 +1,68 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:http/http.dart' as http;
+import 'package:meta/meta.dart';
+
+import '../graphql/ws/websocket_message.dart';
+import 'authorize_http_request.dart';
+
+/// Takes input websocket message (connection or subscription establisher) and
+/// adds authorization headers from auth repo.
+@internal
+Future<WebSocketMessage> authorizeWebSocketMessage(
+  WebSocketMessage inputMessage,
+  AWSApiConfig config,
+  AmplifyAuthProviderRepository authRepo,
+) async {
+  final body = inputMessage.payload?.toJson()['data'];
+  if (inputMessage is WebSocketConnectionInitMessage) {
+    inputMessage.authorizationHeaders =
+        await _generateAuthorizationHeaders(config, authRepo: authRepo);
+  } else if (body is String) {
+    inputMessage.payload?.authorizationHeaders =
+        await _generateAuthorizationHeaders(config,
+            authRepo: authRepo, body: body);
+  }
+  return Future.value(inputMessage);
+}
+
+Future<Map<String, dynamic>> _generateAuthorizationHeaders(
+  AWSApiConfig config, {
+  required AmplifyAuthProviderRepository authRepo,
+  String body = '{}',
+}) async {
+  final endpointHost = Uri.parse(config.endpoint).host;
+  // Create canonical HTTP request to authorize.
+  final maybeConnect = body != '{}' ? '' : '/connect';
+  final canonicalHttpRequest =
+      http.Request('POST', Uri.parse('${config.endpoint}$maybeConnect'));
+  canonicalHttpRequest.headers.addAll({
+    AWSHeaders.accept: 'application/json, text/javascript',
+    AWSHeaders.contentEncoding: 'amz-1.0',
```

Comment:
In fact all of these could just be in a const variable with a link to the AppSync doc that provides these values